File: framework/languages/node/packages/framework/src/contracts/Streams/IZLinkSession.ts
import type { RoutingId, ZLinkMessage } from '../Common';
import type { ZLinkSessionActors } from './IZLinkSessionActor';
import type { ZLinkStreamError } from './ZLinkStreamError';
import type { ZLinkSessionHandlerRegistry } from './IZLinkSessionPacketHandler';

export interface ZLinkSession {
  readonly context: ZLinkSessionContext;
  onConnected?(context: ZLinkSessionContext): Promise<void>;
  onDisconnected?(context: ZLinkSessionContext): Promise<void>;
  onError?(context: ZLinkSessionContext, error: ZLinkStreamError): Promise<void>;
  onDispatch?(dispatch: ZLinkSessionDispatchContext, payload: ZLinkMessage): Promise<void>;
}

export interface ZLinkSessionDispatchContext {
  readonly packetName: string;
  readonly metadata: ReadonlyMap<string, string>;
  readonly canReply: boolean;
}


export interface ZLinkSessionFactory<TSession extends ZLinkSession = ZLinkSession> {
  create(context: ZLinkSessionContext): Promise<TSession>;
}

export interface ZLinkSessionContext {
  readonly sessionId: string;
  readonly routingId?: RoutingId;
  readonly localAddr?: string;
  readonly remoteAddr?: string;
  readonly client: ZLinkSessionClient;
  readonly actors: ZLinkSessionActors;
  readonly handlers: ZLinkSessionHandlerRegistry;
  close(signal?: AbortSignal): Promise<void>;
}

export interface ZLinkSessionClient {
  send(message: unknown): ZLinkSessionSendCall;
  reply(message: unknown): ZLinkSessionReplyCall;
}

export interface ZLinkSessionSendCall {
  metadata(key: string, value: string): this;
  compress(enabled?: boolean): this;
  submit(signal?: AbortSignal): Promise<void>;
}

export interface ZLinkSessionReplyCall {
  compress(enabled?: boolean): this;
  submit(signal?: AbortSignal): Promise<void>;
}

File: framework/languages/dotnet/src/Zlink.Framework/Contracts/Streams/IZLinkSession.cs
namespace Zlink.Framework.Contracts.Streams;

public interface IZLinkSession
{
    IZLinkSessionContext Context { get; }

    void Configure()
    {
    }

    ValueTask OnConnectedAsync(
        CancellationToken cancellationToken);

    ValueTask OnDisconnectedAsync(
        CancellationToken cancellationToken);

    ValueTask OnErrorAsync(
        ZLinkStreamError error,
        CancellationToken cancellationToken);

    /// <summary>
    ///     Handles a framework-owned inbound stream payload.
    /// </summary>
    /// <remarks>
    ///     The payload is a framework <see cref="ZLinkMessage" />. Session code may
    ///     decode it or pass it to framework APIs such as
    ///     <see cref="IZLinkSessionActor.RelayAsync" />.
    /// </remarks>
    ValueTask OnDispatchAsync(
        ZLinkSessionDispatchContext dispatch,
        ZLinkMessage payload,
        CancellationToken cancellationToken)
    {
        _ = dispatch;
        _ = payload;
        _ = cancellationToken;
        return ValueTask.CompletedTask;
    }
}

public interface IZLinkSessionClient
{
    IZLinkSessionSendCall Send<TMessage>(TMessage message);

    IZLinkSessionReplyCall Reply<TMessage>(TMessage message);
}

public interface IZLinkSessionActors
{
    IReadOnlyCollection<IZLinkSessionActor> Bound { get; }

    ValueTask<IZLinkSessionActor> BindAsync(
        ActorRef actor,
        CancellationToken cancellationToken = default);

    ValueTask<IZLinkSessionActor> BindOrGetAsync(
        ActorRef actor,
        CancellationToken cancellationToken = default);

    IZLinkSessionActor? Find(string actorId);
}

public interface IZLinkSessionContext
{
    string SessionId { get; }

    RoutingId? RoutingId { get; }

    string? LocalAddr { get; }

    string? RemoteAddr { get; }

    IZLinkSessionClient Client { get; }

    IZLinkSessionActors Actors { get; }

    IZLinkSessionHandlerRegistry Handlers { get; }

    ValueTask CloseAsync();
}

public interface IZLinkSessionSendCall
    : IZLinkMetadataCall<IZLinkSessionSendCall>
{
    IZLinkSessionSendCall Compress();

    /// <summary>
    /// Submits the session send. Input validation and local transport
    /// acceptance complete before this method returns. If the transport queue
    /// is full, the call waits up to the configured send timeout.
    /// </summary>
    ValueTask Async(
        CancellationToken cancellationToken = default);
}

public interface IZLinkSessionReplyCall
{
    IZLinkSessionReplyCall Compress();

    /// <summary>
    /// Submits a reply for the currently handled request packet. The reply uses
    /// the request packet name and does not expose PacketName because a reply
    /// must keep the request correlation.
    /// </summary>
    ValueTask Async(
        CancellationToken cancellationToken = default);
}
